Lib/ufo2ft/filters/__init__.py

```diff
@@ -53,176 +79,71 @@ def loadFilters(ufo):
-                if glyphName in modified:
-                    continue
-                glyph = glyphSet[glyphName]
-                if include(glyph) and filter_(glyph):
-                    modified.add(glyphName)
-
-        num = len(modified)
-        if num > 0:
-            logger.debug(
-                "Took %.3fs to run %s on %d glyph%s",
-                t,
-                self.name,
-                len(modified),
-                "" if num == 1 else "s",
-            )
-        return modified
+_filterSpecRE = re.compile(
+    r"(?:([\w\.]+)::)?"  # MODULE_NAME + '::'
+    r"(\w+)"  # CLASS_NAME [required]
+    r"(?:\((.*)\))?"  # (KWARGS)
+)
+
+
+def isValidFilter(klass):
+    """Return True if 'klass' is a valid filter class.
+    A valid filter class is a class (of type 'type'), that has
+    a '__call__' (bound method), with the signature matching the same method
+    from the BaseFilter class:
+
+           def __call__(self, font, feaFile, compiler=None)
+    """
+    if not isclass(klass):
+        logger.error("{klass!r} is not a class")
+        return False
+    if not callable(klass):
+        logger.error("{klass!r} is not callable")
+        return False
+    if getfullargspec(klass.__call__).args != getfullargspec(BaseFilter.__call__).args:
+        logger.error("{klass!r} '__call__' method has incorrect signature", klass)
+        return False
+    return True
+
+
+def loadFilterFromString(spec):
+    """Take a string specifying a filter class to load (either a built-in
+    writer or one defined in an external, user-defined module), initialize it
+    with given options and return the filter object.
+
+    The string must conform to the following notation:
+    - an optional python module, followed by '::'
+    - a required class name; the class must have a method called 'filter'
+      with the same signature as the BaseFilter.
+    - an optional list of keyword-only arguments enclosed by parentheses
+
+    Raises ValueError if the string doesn't conform to this specification;
+    TypeError if imported name is not a filter class; and ImportError if the
+    user-defined module cannot be imported.
+
+    Examples:
+
+    >>> loadFilterFromString("ufo2ft.filters.removeOverlaps::RemoveOverlapsFilter")
+    <ufo2ft.filters.removeOverlaps.RemoveOverlapsFilter object at ...>
+    """
+    spec = spec.strip()
+    m = _filterSpecRE.match(spec)
+    if not m or (m.end() - m.start()) != len(spec):
+        raise ValueError(spec)
+    moduleName = m.group(1) or "ufo2ft.filters"
+    className = m.group(2)
+    kwargs = m.group(3)
+
+    module = importlib.import_module(moduleName)
+    klass = getattr(module, className)
+    if not isValidFilter(klass):
+        raise TypeError(klass)
+    try:
+        options = _kwargsEval(kwargs) if kwargs else {}
+    except SyntaxError:
+        raise ValueError("options have incorrect format: %r" % kwargs)
+    if "args" in options:
+        args = options.pop("args")
```

Comment:
hm I don't like this hack to use a keyword named `args` for the positional init arguments. Is that because we only evaluate the keywords? Can't we just only support keyword initialization? Or do some filters require positional-only arguments that can't also be expressed as kwargs?

Reply:
I see you wrote "None of the ufo2ft filters use the args.", so why you handling them here?

Reply:
oh I see, the base filter constructor doesn't merge the args and kwargs, but treat each independently. Maybe we should fix that, and allow to specify positional arguments that are listed in the `_args` as keyword parameters as well. Then in the string definition, we only suppot eval'ing keywords only.

Reply:
Sorry I changed the commit to keep args after all. I should have updated the original description. The string definition allows having args in the dictionary with the key "args".

Reply:
> The string definition allows having args in the dictionary with the key "args".

yes, and I think that's weird. We should not need to support that.

Reply:
OK. I'll have a go.